### PR TITLE
DACT-1819 Refactor address validation into a single class

### PR DESCRIPTION
**Pull request by jditchburn**
Refactor address validation so that residential and correspondence addresses both run through the same validation code.
https://companieshouse.atlassian.net/browse/DACT-1819

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/officerfiling/api/validation/AddressValidator.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    private final List<String> ukCountryList;
+
+    protected AddressValidator(Logger logger, CompanyProfileService companyProfileService, String inputAllowedNationalities, ApiEnumerations apiEnumerations, List<String> countryList, List<String> ukCountryList) {
+        super(logger, companyProfileService, inputAllowedNationalities, apiEnumerations);
+        this.countryList = countryList;
+        this.ukCountryList = ukCountryList;
+    }
+
+    @Override
+    public void validateRequiredDtoFields(HttpServletRequest request, List<ApiError> errorList, OfficerFilingDto dto) {
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/officerfiling/api/validation/AddressValidator.java:35]: just mark method abstract?

File: src/main/java/uk/gov/companieshouse/officerfiling/api/validation/AddressValidator.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+            if (!validateDtoFieldLength(addressLineOne, LENGTH_50)) {
+                createValidationError(request, errorList, addressErrorProvider.getAddressLineOneLength());
+            }
+            if (!isValidCharacters(addressLineOne)) {
+                createValidationError(request, errorList, addressErrorProvider.getAddressLineOneCharacters());
+            }
+        }
+    }
+
+    private void validateAddressLineTwo(AddressErrorProvider addressErrorProvider, HttpServletRequest request, List<ApiError> errorList, String addressLineTwo) {
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/officerfiling/api/validation/AddressValidator.java:100]: Would it be helpfull to say Optional in the method name? Same for region.

File: src/main/java/uk/gov/companieshouse/officerfiling/api/validation/AddressValidator.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    private void createBlankValidationErrors(AddressErrorProvider addressErrorProvider, HttpServletRequest request, List<ApiError> errorList) {
+        createValidationError(request, errorList, addressErrorProvider.getPremisesBlank());
+        createValidationError(request, errorList, addressErrorProvider.getAddressLineOneBlank());
+        createValidationError(request, errorList, addressErrorProvider.getLocalityBlank());
+        createValidationError(request, errorList, addressErrorProvider.getPostalCodeBlank());
+        createValidationError(request, errorList, addressErrorProvider.getCountryBlank());
+    }
+
+    private void validatePremises(AddressErrorProvider addressErrorProvider, HttpServletRequest request, List<ApiError> errorList, String premises) {
+        if (premises == null || premises.isBlank()) {
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/officerfiling/api/validation/AddressValidator.java:75]: You could use if(StringUtils.isBlank(premises)) 
that avoid the null check and cuts down testing.

File: src/test/java/uk/gov/companieshouse/officerfiling/api/validation/OfficerAppointmentValidatorTest.java [hunk 4]
```diff
@@ -1827,7 +1827,7 @@ void validateAppointmentDateAfterIncorporationDateWhenValid() {
                 .referenceAppointmentId(FILING_ID)
                 .appointedOn(LocalDate.of(2023, Month.JANUARY, 5))
                 .build();
-        officerAppointmentValidator.validateAppointmentDateBeforeIncorporationDate(request, apiErrorsList, officerFilingDto, companyProfile);
```
Comment (AndreKramer7654) [inline src/test/java/uk/gov/companieshouse/officerfiling/api/validation/OfficerAppointmentValidatorTest.java:1830]: Are you changing the name back? I think Milan renamed it validateAppointmentDateBeforeIncorporationDate

File: src/test/java/uk/gov/companieshouse/officerfiling/api/validation/AddressValidatorTest.java [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+    @BeforeEach
+    void setUp() {
+        addressValidator = new AddressValidator(mockLogger, mockCompanyProfileService, ALLOWED_NATIONALITIES, mockApiEnumerations, COUNTRY_LIST, UK_COUNTRY_LIST);
+        apiErrorList = new ArrayList<>();
+        residentialAddressErrorProvider = new ResidentialAddressErrorProvider(mockApiEnumerations);
+        correspondenceAddressErrorProvider = new CorrespondenceAddressErrorProvider(mockApiEnumerations);
+    }
+
+    @Test
+    void validateWhenAddressFieldsAreValid() {
```
Comment (AndreKramer7654) [inline src/test/java/uk/gov/companieshouse/officerfiling/api/validation/AddressValidatorTest.java:57]: Add test for optional line2 and region being null (or blank)?

File: src/test/java/uk/gov/companieshouse/officerfiling/api/validation/AddressValidatorTest.java [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+
+        assertThat(apiErrorList)
+                .as("No errors should be produced when all fields are valid")
+                .hasSize(1)
+                .extracting(ApiError::getError)
+                .contains("Country Invalid");
+    }
+
+    @Test
+    void validateWhenPostCodeIsUkInvalid() {
```
Comment (AndreKramer7654) [inline src/test/java/uk/gov/companieshouse/officerfiling/api/validation/AddressValidatorTest.java:288]: IsUkInvalid?

File: src/main/java/uk/gov/companieshouse/officerfiling/api/validation/AddressValidator.java (hunk 1 above)
Comment (jditchburn) [reply src/main/java/uk/gov/companieshouse/officerfiling/api/validation/AddressValidator.java:35]: But then I would have to make the class abstract? I thought that would be confusing as we're not expecting anything to extend this class

Comment (AndreKramer7654) [reply src/main/java/uk/gov/companieshouse/officerfiling/api/validation/AddressValidator.java:36]: It should not extend OfficerValidator. Then you don't need to Override.

Comment (AndreKramer7654) [reply src/main/java/uk/gov/companieshouse/officerfiling/api/validation/AddressValidator.java:36]: But then you need access to quite a few methods in OfficerValidator. Maybe not throwing an exception is the better approach if there are no fields to validate. We could validate the two addresses via these methods but I don't think that's right either. I'd prefer not to have the NotImplemented exceptions over having empty methods.

Comment (jditchburn) [reply src/main/java/uk/gov/companieshouse/officerfiling/api/validation/AddressValidator.java:36]: But OfficerValidator currently contains a lot of validation methods that are used in the address validation. I'd rather include these two overridden methods than duplicate all those methods in this class. Those methods could be moved somewhere else to be re-used across all the validators but that would require refactoring quite a lot of code in the other validators.

Comment (jditchburn) [reply src/main/java/uk/gov/companieshouse/officerfiling/api/validation/AddressValidator.java:36]: To be honest I'd prefer to just remove them from the abstract class as they're not even all implemented in the other validators.

Comment (AndreKramer7654) [reply src/main/java/uk/gov/companieshouse/officerfiling/api/validation/AddressValidator.java:36]: That sounds the best plan thanks.

**Review by AndreKramer7654 — APPROVED** [review]

**Review by PhilipGlover — APPROVED** [review]
- refactoring looks good
- I may be missing it, but shouldn't the `OfficerUpdateValidator` also call `addressValidator.validate` if the address fields have been changed?  Is that another story?

**Comment by jditchburn** [comment]
> * refactoring looks good
> * I may be missing it, but shouldn't the `OfficerUpdateValidator` also call `addressValidator.validate` if the address fields have been changed?  Is that another story?

Yes it just hasn't been implemented yet. We'll be adding address validation to OfficerUpdateValidator this sprint after this PR is merged.
